extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/UserData.java

```diff
@@ -0,0 +1,59 @@
+package io.quarkus.websockets.next;
+
+/**
+ * Mutable user data associated with a connection. Implementations must be thread-safe.
+ */
+public interface UserData {
+
+    /**
+     *
+     * @param <VALUE>
+     * @param key
+     * @return the value or {@code null} if no mapping is found
+     */
+    <VALUE> VALUE get(TypedKey<VALUE> key);
+
+    /**
+     * Associates the specified value with the specified key. An old value is replaced by the specified value.
+     *
+     * @param <ConnectionData.VALUE>
+     * @param key
+     * @param value
+     * @return the previous value associated with {@code key}, or {@code null} if no mapping exists
+     */
+    <VALUE> VALUE put(TypedKey<VALUE> key, VALUE value);
+
+    /**
+     *
+     * @param <VALUE>
+     * @param key
+     */
+    <VALUE> VALUE remove(TypedKey<VALUE> key);
+
+    int size();
+
+    void clear();
+
+    /**
+     * @param <TYPE> The type this key is used for.
+     */
+    record TypedKey<TYPE>(String value) {
```

Comment:
You don't really need the string key. The object identity of the key is itself good enough.

This

```java
static class Key<T> {
}
```

is all you need.

Reply:
Not really. This way you would need to use the object reference as a key. Always. That's not what we want/need. A string-based key is IMO more practical.

Reply:
Well, you're risking collisions, so the string keys will have to be namespaced... Not a fan. But your choice.